crypto/signatures.go

```diff
@@ -23,25 +23,58 @@ type (
 var (
 	ErrNilInput         = errors.New("cannot use nil input")
 	ErrInvalidSignature = errors.New("invalid signature")
+	ErrRandUnexpected   = errors.New("unexpected result from random number generator")
 )
 
-// GenerateKeyPair creates a public-secret keypair that can be used to sign and
-// verify messages.
-func GenerateSignatureKeys() (sk SecretKey, pk PublicKey, err error) {
+// KeyPairGenerator is an interface that allows the caller to generate
+// public-secret key pairs.
+type KeyPairGenerator interface {
+	Generate() (SecretKey, PublicKey, error)
+	GenerateDetermistically(entropy [EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// readBytesFunc is a function pointer that reads bytes into a buffer and
+// returns the total number of bytes written to the buffer.
+type readBytesFunc func([]byte) (int, error)
+
+// deriveEd25519KeyPairFunc is a function pointer that matches the signature of
+// ed25519.GenerateKey.
+type deriveEd25519KeyPairFunc func([EntropySize]byte) (ed25519.SecretKey, ed25519.PublicKey)
+
+// SignatureKeyGenerator is an implementation of KeyPairGenerator.
+type SignatureKeyGenerator struct {
+	readRandBytes readBytesFunc
+	deriveKeyPair deriveEd25519KeyPairFunc
+}
+
+// NewSignatureKeyGenerator creates a new SignatureKeyGenerator type that uses
+// random data and depends on the ed25519.GenerateKey function for deriving
+// key pairs.
+func NewSignatureKeyGenerator() SignatureKeyGenerator {
+	return SignatureKeyGenerator{rand.Read, ed25519.GenerateKey}
+}
+
+// Generate creates a public-secret keypair that can be used to sign and verify
+// messages.
+func (skg SignatureKeyGenerator) Generate() (sk SecretKey, pk PublicKey, err error) {
 	var entropy [EntropySize]byte
-	_, err = rand.Read(entropy[:])
+	written, err := skg.readRandBytes(entropy[:])
 	if err != nil {
 		return
 	}
+	if written != EntropySize {
+		// readRandBytes did not fill the buffer. This should never happen.
+		return sk, pk, ErrRandUnexpected
```

Comment:
We can be flexible if we want to. I was being especially rigorous in this PR about checking every possible behavior of `rand.Read` to demonstrate the value of breaking up dependencies and the increased coverage we can get that we can't get with other types of testing.

I don't know that I'd do this normally, given that it's unlikely we'll ever use any implementation other than `rand.Read` which provides this guarantee.